File: LeetCode-75/198-M-House-Robber/house-robber.py
class Solution:
    def rob(self, nums: list[int]) -> int:
        length = len(nums)
        if length == 0:
            return -99 # empty list
        if length == 1:
            return nums[0]
        
        prev2 = nums[0]
        prev1 = max(nums[0], nums[1])
        curr = 0
        
        for i in range(2, length):
            curr = max(nums[i]+prev2, prev1)
            prev2 = prev1
            prev1 = curr
        return prev1

File: LeetCode-75/198-M-House-Robber/house-robber.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        if (nums.empty()) return -99; // invalid input.
        int len = nums.size();
        if (len==1) return nums[0];

        int prev2 = nums[0];
        int prev1 = max(nums[0], nums[1]);
        int curr = 0;
        for (int k=2; k<len; k++)
        {
            curr = max(nums[k]+prev2, prev1);
            prev2 = prev1;
            prev1 = curr;
        }
        return prev1;
    }
};